pkg/crc/pullsecret/pullsecret.go

```diff
@@ -0,0 +1,169 @@
+	// the pod are not completely stopped and when remove happen it will throw
+	// an error like below.
+	// remove /var/run/containers/storage/overlay-containers/97e5858e610afc9f71d145b1a7bd5ad930e537ccae79969ae256636f7fb7e77c/userdata/shm: device or resource busy
+	stopAndRemovePodsCmd = `bash -c 'sudo crictl stopp $(sudo crictl pods -q) && sleep 2 &&\
+sudo crictl rmp $(sudo crictl pods -q)'`
+)
+
+func AddPullSecretAndClusterID(driver drivers.Driver, pullSec string, kubeconfigFilePath string) error {
+	// Add the kubeconfig File to the Instance.
+	if err := addKubeconfigFileToInstance(driver, kubeconfigFilePath); err != nil {
+		return err
+	}
+
+	// Add the Pull secret kubernetes resource definition to Instance
+	if err := addpullSecretSpecToInstance(driver, pullSec); err != nil {
+		return err
+	}
+
+	// Replace user pull secret and add cluster ID
+	if err := setPullSecretAndClusterID(driver); err != nil {
+		return err
+	}
+
+	// Add user pull secret to the instance
+	if err := addpullSecretToIntanceDisk(driver, pullSec); err != nil {
+		return err
+	}
+	return nil
+}
+
+func addpullSecretSpecToInstance(driver drivers.Driver, pullSec string) error {
+	base64OfPullSec := base64.StdEncoding.EncodeToString([]byte(pullSec))
+	output, err := drivers.RunSSHCommandFromDriver(driver, fmt.Sprintf("cat <<EOF | tee /tmp/pull-secret.yaml\n%s\nEOF", fmt.Sprintf(pullSecret, base64OfPullSec)))
+	if err != nil {
+		return err
+	}
+	logging.DebugF("Output is : %s", output)
+	return nil
+}
+
+func addpullSecretToIntanceDisk(driver drivers.Driver, pullSec string) error {
+	output, err := drivers.RunSSHCommandFromDriver(driver, fmt.Sprintf("cat <<EOF | sudo tee /var/lib/kubelet/config.json\n%s\nEOF", pullSec))
+	if err != nil {
+		return err
+	}
+	logging.DebugF("Output is : %s", output)
+	return nil
+}
+
+func addKubeconfigFileToInstance(driver drivers.Driver, kubeconfigFilePath string) error {
+	kubeconfig, err := ioutil.ReadFile(kubeconfigFilePath)
+	if err != nil {
+		return err
+	}
+	_, err = drivers.RunSSHCommandFromDriver(driver, fmt.Sprintf("cat <<EOF | tee /tmp/kubeconfig\n%s\nEOF", string(kubeconfig)))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func replaceUserPullSecret(driver drivers.Driver) error {
+	output, err := drivers.RunSSHCommandFromDriver(driver, replacePullSecretCmd)
+	if err != nil {
+		return err
+	}
+	logging.DebugF("Output of %s: %s", replacePullSecretCmd, output)
+	return nil
+}
+
+func addClusterID(driver drivers.Driver) error {
+	clusterID := uuid.New()
+	output, err := drivers.RunSSHCommandFromDriver(driver, fmt.Sprintf(updateClusterIdCmd, clusterID))
+	if err != nil {
+		return err
+	}
+	logging.DebugF("Output of %s: %s", updateClusterIdCmd, output)
+	return nil
+}
+
+func setPullSecretAndClusterID(driver drivers.Driver) (rerr error) {
+	sd := systemd.NewInstanceSystemdCommander(driver)
+	defer func() {
+		// Stop the kublet service.
+		if _, err := sd.Stop("kubelet"); err != nil {
+			rerr = err
+		}
```

Comment:
If `sd.Start("kubelet");` fails, it's likely `sd.Stop("kubelet")` would fail too, and if I'm not mistaken, we will overwrite `setPullSecretAndCLusterId` return value with this error, which I don't think is desirable. Maybe change `if err != nil { rerr = err }` to `if rerr == nil { rerr = err }`

Reply:
@cfergeau then what if the start happen successful but stop failed then how would we know?

Reply:
If everything was successful until the `defer`, then `rerr` will be `nil`. If then something fails in the `defer`, then we'll overwrite the `nil` `rerr` with the error that happened. If something failed before, then we will not be overwriting it.

Reply:
@cfergeau Ah, I got it now, so I am going to use the `multierror` functionality.

Reply:
Hmm, why not, but I think once the first error happened, we can ignore the ones which happen on cleanup.